### PR TITLE
Add CallableDiscriminator and Tag

### PR DESCRIPTION
**Pull request by dmontagu**
I think this should be preferred over https://github.com/pydantic/pydantic/pull/6915, mostly because the implementation there involves duplicating the list of types in the union, and generally seems more fragile.

I've opened as draft until docs/tests are added, hoping @sydney-runkle can help get this across the line tomorrow.

### CODE REVIEW

File: pydantic/types.py [hunk 1]
```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
...
+        if original_schema['type'] != 'union':
+            # This likely indicates that the schema was a single-item union that was simplified.
+            # In this case, we do the same thing we do in
+            # `pydantic._internal._discriminated_union._ApplyInferredDiscriminator._apply_to_root`, namely,
+            # package the generated schema back into a single-item union.
+            original_schema = core_schema.union_schema([original_schema])
+
+        tagged_union_choices = {}
+        for i, choice in enumerate(original_schema['choices']):
+            tag = f'case-{i}'
```
Comment (dmontagu) [inline pydantic/types.py:2494]: You can kind of see what the default tag selection logic is here.

This might merit some reconsideration; either way we probably want to add some tests.

Another option which might be better would be to just error if the tag isn't explicitly specified. (I.e., if the `metadata.get(_TAGGED_UNION_TAG_KEY)` returns `None` below).

File: pydantic/types.py [hunk 2]
```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
                 return object.__getattribute__(self, item)
 
     __hash__ = object.__hash__
+
+
+_TAGGED_UNION_TAG_KEY = 'pydantic-tagged-union-tag'
```
Comment (dmontagu) [inline pydantic/types.py:2442]: This should probably be moved to `pydantic._internal._core_utils`, like `NEEDS_APPLY_DISCRIMINATED_UNION_METADATA_KEY`, `HAS_INVALID_SCHEMAS_METADATA_KEY`, etc.

File: pydantic/types.py [hunk 3]
```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
...
+
+
+@_dataclasses.dataclass(**_internal_dataclass.slots_true, frozen=True)
+class Tag:
```
Comment (dmontagu) [inline pydantic/types.py:2446]: This can be renamed if appropriate.

File: pydantic/types.py [hunk 4]
```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
                 return object.__getattribute__(self, item)
 
     __hash__ = object.__hash__
+
+
+_TAGGED_UNION_TAG_KEY = 'pydantic-tagged-union-tag'
+
+
+@_dataclasses.dataclass(**_internal_dataclass.slots_true, frozen=True)
```
Comment (dmontagu) [inline pydantic/types.py:2445]: `frozen=True` is used here because annotations should be hashable, or you run into issues with caching generic containers (such as `typing.Union`!).

Technically we only need hashability, and we could get that with just `__hash__ = object.__hash__`, but looking at the fields/usage of both this and `CallableDiscriminator`, I didn't see any reason these types shouldn't just be frozen anyway. So that kept things simple.

File: pydantic/types.py [hunk 5]
```diff
@@ -2429,3 +2437,92 @@ def __getattr__(self, item: str) -> Any:
...
+class Tag:
+    """Provides a way to specify the expected tag to use for a case with a callable discriminated union.
+
+    TODO: Need to add more thorough docs..
```
Comment (dmontagu) [inline pydantic/types.py:2449]: ☝️

File: pydantic/types.py (hunk 3 above)
Comment (sydney-runkle) [reply pydantic/types.py:2446]: I like `Tag` 👍

File: pydantic/types.py (hunk 2 above)
Comment (sydney-runkle) [reply pydantic/types.py]: Resolving, as I've moved this as requested

File: pydantic/types.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/types.py]: I removed the support for default tags for each choice and added an associated error for when a tag is missing for a case. I added docs + tests as well 👍

File: pydantic/_internal/_core_utils.py [hunk 6]
```diff
@@ -48,6 +48,10 @@
 schema building because one of it's members refers to a definition that was not yet defined when the union
 was first encountered.
 """
+TAGGED_UNION_TAG_KEY = 'pydantic-tagged-union-tag'
```
Comment (dmontagu) [inline pydantic/_internal/_core_utils.py:51]: Let's change this value to match format of the others surrounding it:
```suggestion
TAGGED_UNION_TAG_KEY = 'pydantic.internal.tag'
```

File: docs/api/standard_library_types.md [hunk 7]
```diff
@@ -871,18 +871,24 @@ print(type(User(id='123', age='45').id))
...
+models and primitive types.
 
 Setting a discriminated union has many benefits:
 
 - validation is faster since it is only attempted against one model
 - only one explicit error is raised in case of failure
 - the generated JSON schema implements the [associated OpenAPI specification](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#discriminator-object)
 
+#### Discriminated Unions with `str` discriminators
+In the case of a `Union` with multiple models, you sometimes know exactly which field of each
```
Comment (dmontagu) [inline docs/api/standard_library_types.md:885]: I would maybe rephrase/add a bit of context here:

"Frequently, there is a common field to all members of the union that can be used to distinguish which union case the data should be validated against; this is referred to as the "discriminator" in [OpenAPI](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/)."

Can ignore this if you prefer current phrasing/discussion.

File: docs/api/standard_library_types.md [hunk 8]
```diff
@@ -922,9 +928,125 @@ except ValidationError as e:
...
+    For `str` discriminators:
+        * `some_field: Union[...] = Field(discriminator='my_discriminator'`
+        * `some_field: Annotated[Union[...], Field(discriminator='my_discriminator')]`
+    For `CallableDiscriminator` discriminators:
+        * `some_field: Union[...] = Field(discriminator=CallableDiscriminator(...))`
+        * `some_field: Annotated[Union[...], CallableDiscriminator(...)]`
+        * `some_field: Annotated[Union[...], Field(discriminator=CallableDiscriminator(...))]`
 
 !!! warning
     Discriminated unions cannot be used with only a single variant, such as `Union[Cat]`.
```
Comment (dmontagu) [inline docs/api/standard_library_types.md:1052]: Not sure if this applies to callable discriminators. And there's a non-contrived (imo, anyway) use case for this — faster failing if a specific key/value pair is missing. I mean, it will rarely be relevant, but you could imagine an API where you have many object types but you only want to allow one variant in a certain position, and don't need to validate/produce a complex error message if the tag is just wrong.

I don't think we need to necessarily document that usage pattern, but I'm just thinking it maybe makes sense to update this warning given that; I defer to you @sydney-runkle. (I also don't promise it actually works with just a single case, I just think it does based on the implementation 😄)

File: docs/concepts/fields.md [hunk 9]
```diff
@@ -667,7 +667,8 @@ print(repr(Model.model_validate({'pet': {'pet_kind': 'dog', 'age': 12}})))
 #> Model(pet=Dog(pet_kind='dog', age=12))
 ```
 
-See the [Discriminated Unions] for more details.
+You can also take advantage of `Annotated` to define your discriminated unions.
+See the [Discriminated Unions] docs for more details.
```
Comment (dmontagu) [inline docs/concepts/fields.md:671]: Does this need a link (like, `[Discriminated Unions](#discriminated_unions)` or whatever)? Or is there something magical happening with it auto-generating links from just square brackets? I recognize it was like this in current docs but seems like it's worth fixing if it's just a broken link now.

Comment (dmontagu) [reply docs/concepts/fields.md:671]: Ah I see it can be specified at the bottom, nevermind.

File: docs/concepts/performance.md [hunk 10]
```diff
@@ -207,3 +209,5 @@ Instead of using nested models, use `TypedDict` to define the structure of the d
 ## Avoid wrap validators if you really care about performance
 
 <!-- TODO: I need help on this one. -->
```
Comment (dmontagu) [inline docs/concepts/performance.md:211]: lol

File: docs/concepts/fields.md (hunk 9 above)
Comment (sydney-runkle) [reply docs/concepts/fields.md:671]: The link is at the bottom of the page :)

File: docs/errors/usage_errors.md [hunk 11]
```diff
@@ -365,6 +365,62 @@ class Model(BaseModel):
 assert Model(pet={'pet_type': 'kitten'}).pet.pet_type == 'cat'
 ```
 
+## Callable discriminator no tag {#callable-discriminator-no-tag}
```
Comment (dmontagu) [inline docs/errors/usage_errors.md:368]:
```suggestion
## Callable discriminator case with no tag {#callable-discriminator-no-tag}
```
The current name of this section feels a little too far from a sensible english statement. Looking briefly at some of the other section titles, even if the title isn't a viable sentence, seem to be like, a more free-standing-capable phrase, if that makes sense.

File: pydantic/types.py [hunk 12]
```diff
@@ -2429,3 +2438,216 @@ def __getattr__(self, item: str) -> Any:
...
+            # In this case, we do the same thing we do in
+            # `pydantic._internal._discriminated_union._ApplyInferredDiscriminator._apply_to_root`, namely,
+            # package the generated schema back into a single-item union.
+            original_schema = core_schema.union_schema([original_schema])
+
+        tagged_union_choices = {}
+        for i, choice in enumerate(original_schema['choices']):
+            tag = None
+            if isinstance(choice, tuple):
+                choice, tag = choice
```
Comment (dmontagu) [inline pydantic/types.py:2616]: Let's just make sure that this doesn't cause us any issues with "default" tags. I can't remember if we exposed this on pydantic side or just pydantic-core currently..

File: docs/api/standard_library_types.md (hunk 8 above)
Comment (sydney-runkle) [reply docs/api/standard_library_types.md:1057]: @dmontagu,

I think the warning still holds:

```py
from typing import Any, Literal, Union

from typing_extensions import Annotated

from pydantic import BaseModel, TypeAdapter, CallableDiscriminator, Tag


def get_discriminator_value(v: Any) -> str:
    if isinstance(v, dict):
        return v.get('pet_type')
    return getattr(v, 'fruit', None)


class Cat(BaseModel):
    pet_type: Literal['cat'] = 'cat'


ta = TypeAdapter(Annotated[Union[Annotated[Cat, Tag('cat')]], CallableDiscriminator(get_discriminator_value)])
#> CallableDiscriminator must be used with a Union type, not <class '__main__.Cat'>
```

Comment (sydney-runkle) [reply docs/api/standard_library_types.md:1057]: Resolving + leaving the warning in place 😢

File: docs/api/standard_library_types.md (hunk 7 above)
Comment (sydney-runkle) [reply docs/api/standard_library_types.md:885]: I like your suggestion. Thanks for the feedback. I changed it, so marking as resolved

File: pydantic/types.py (hunk 12 above)
Comment (dmontagu) [reply pydantic/types.py:2624]: investigated, and I think we can use Tag to set this as a way to improve other usability of unions in pydantic, but I think we can ignore that in this PR.

**Review by sydney-runkle — APPROVED** [review]
Looks good, thanks for the fast turnaround @dmontagu!
